ethash.go

```diff
@@ -38,279 +45,177 @@ import (
-	Epoch  uint64
-}
-
-type DAG struct {
-	dag            unsafe.Pointer // full GB of memory for dag
-	file           bool
-	paramsAndCache *ParamsAndCache
-}
-
-type Ethash struct {
-	turbo          bool
-	HashRate       int64
-	chainManager   pow.ChainManager
-	dag            *DAG
-	paramsAndCache *ParamsAndCache
-	ret            *C.ethash_return_value
-	dagMutex       *sync.RWMutex
-	cacheMutex     *sync.RWMutex
-}
-
-func parseNonce(nonce []byte) (uint64, error) {
-	nonceBuf := bytes.NewBuffer(nonce)
-	nonceInt, err := binary.ReadUvarint(nonceBuf)
-	if err != nil {
-		return 0, err
-	}
-	return nonceInt, nil
-}
-
 const epochLength uint64 = 30000
 
-func makeParamsAndCache(chainManager pow.ChainManager, blockNum uint64) (*ParamsAndCache, error) {
-	if blockNum >= epochLength*2048 {
-		return nil, fmt.Errorf("block number is out of bounds (value %v, limit is %v)", blockNum, epochLength*2048)
-	}
-	paramsAndCache := &ParamsAndCache{
-		params: new(C.ethash_params),
-		cache:  new(C.ethash_cache),
-		Epoch:  blockNum / epochLength,
-	}
-	C.ethash_params_init(paramsAndCache.params, C.uint32_t(uint32(blockNum)))
-	paramsAndCache.cache.mem = C.malloc(C.size_t(paramsAndCache.params.cache_size))
+var minDifficulty = new(big.Int).Exp(big.NewInt(2), big.NewInt(256), big.NewInt(0))
 
-	seedHash, err := GetSeedHash(blockNum)
-	if err != nil {
-		return nil, err
-	}
+var (
+	DefaultDir string = defaultDir()
+	TheLight   Light
+)
 
-	glog.V(logger.Info).Infof("Making cache for epoch: %d (%v) (%x)\n", paramsAndCache.Epoch, blockNum, seedHash)
-	start := time.Now()
-	C.ethash_mkcache(paramsAndCache.cache, paramsAndCache.params, (*C.ethash_h256_t)(unsafe.Pointer(&seedHash[0])))
+const (
+	cacheSizeForTesting = C.uint64_t(1024)
+	dagSizeForTesting   = C.uint64_t(1024 * 32)
+)
 
-	if glog.V(logger.Info) {
-		glog.Infoln("Took:", time.Since(start))
+func defaultDir() string {
+	home := os.Getenv("HOME")
+	if user, err := user.Current(); err == nil {
+		home = user.HomeDir
 	}
+	if runtime.GOOS == "windows" {
+		return filepath.Join(home, "AppData", "Ethash")
+	}
+	return filepath.Join(home, ".ethash")
+}
 
-	return paramsAndCache, nil
+type cache struct {
+	light *C.struct_ethash_light
 }
 
-func (pow *Ethash) UpdateCache(blockNum uint64, force bool) error {
-	pow.cacheMutex.Lock()
-	defer pow.cacheMutex.Unlock()
+func freeCache(h *cache) {
+	C.ethash_light_delete(h.light)
+}
 
-	thisEpoch := blockNum / epochLength
-	if force || pow.paramsAndCache.Epoch != thisEpoch {
-		var err error
-		pow.paramsAndCache, err = makeParamsAndCache(pow.chainManager, blockNum)
-		if err != nil {
-			panic(err)
-		}
+func makeCache(blockNum uint64, test bool) *cache {
+	seedHash, _ := GetSeedHash(blockNum)
+	size := C.ethash_get_cachesize(C.uint64_t(blockNum))
+	if test {
+		size = cacheSizeForTesting
 	}
-
-	return nil
+	light := C.ethash_light_new(size, (*C.ethash_h256_t)(unsafe.Pointer(&seedHash[0])))
+	cache := &cache{light}
+	runtime.SetFinalizer(cache, freeCache)
```

Comment:
We also use finalizers for the DAG. The reasoning is that the mmapped DAG does not count toward real memory usage, so it should be fine to have more than one mapped at the same time. Note that this might not work so well for ‎32bit platforms.                                                                                                                                                                                                                                                                                                                                                                                      ‎